src/java.base/share/classes/sun/security/provider/ML_DSA.java

```diff
@@ -0,0 +1,1427 @@
+        if (sig.hint() == null) {return false;}
+
+        //Decode pk
+        ML_DSA_PublicKey pk = pkDecode(pkBytes);
+
+        //Expand A
+        int[][][] aHat = generateA(pk.rho());
+
+        //Generate tr
+        hash.update(pkBytes);
+        byte[] tr = hash.squeeze(mlDsaTrLength);
+        hash.reset();
+
+        //Generate mu
+        hash.update(tr);
+        hash.update(message);
+        byte[] mu = hash.squeeze(mlDsaMuLength);
+        hash.reset();
+
+        //Get verifiers challenge
+        int[] cHat = sampleInBall(sig.commitmentHash());
+        mlDsaNtt(cHat);
+
+        //Compute response norm and put it in NTT domain
+        boolean zNorm = vectorNormBound(sig.response(), gamma1 - beta);
+        mlDsaVectorNtt(sig.response());
+
+        //Reconstruct signer's commitment
+        int[][] aHatZ = matrixVectorPointwiseMultiply(aHat, sig.response());
+        int[][] t1Hat = vectorConstMul(1 << mlDsa_d, pk.t1());
+        mlDsaVectorNtt(t1Hat);
+        int[][] wApprox = vectorSub(aHatZ,nttConstMultiply(cHat, t1Hat));
+        mlDsaVectorInverseNtt(wApprox);
+        int[][] w1Prime = useHint(sig.hint(), wApprox);
+
+        //Hash signer's commitment
+        hash.update(mu);
+        hash.update(simpleBitPack(wCoeffSize, w1Prime));
+        byte[] cTildePrime = hash.squeeze(lambda/4);
+
+        //Check verify conditions
+        boolean hashEq = Arrays.equals(sig.commitmentHash(), cTildePrime);
+        return !zNorm && hashEq;
+    }
+
+    /*
+    Data conversion functions in Section 7.1 of specification
+     */
+
+    // Bit-pack the t1 and w1 vector into a byte array.
+    // The coefficients of the polynomials in the vector should be nonnegative and less than 2^bitsPerCoeff .
+    public byte[] simpleBitPack(int bitsPerCoeff, int[][] vector) {
+        byte[] result = new byte[(mlDsa_k * mlDsa_n * bitsPerCoeff) / 8];
+        int acc = 0;
+        int shift = 0;
+        int i = 0;
+        for (int[] poly : vector) {
+            for (int m = 0; m < mlDsa_n; m++) {
+                acc += (poly[m] << shift);
+                shift += bitsPerCoeff;
+                while (shift >= 8) {
+                    result[i++] = (byte) acc;
+                    acc >>= 8;
+                    shift -= 8;
+                }
+            }
+        }
+        return result;
+    }
+
+    public byte[] bitPack(int[][] vector, int bitsPerCoeff, int maxValue) {
+        int vecLen = vector.length;
+        byte[] result = new byte[(vecLen * mlDsa_n * bitsPerCoeff) / 8];
+        int acc = 0;
+        int shift = 0;
+        int i = 0;
+        for (int[] poly : vector) {
+            for (int m = 0; m < mlDsa_n; m++) {
+                acc += (maxValue - poly[m]) << shift;
+                shift += bitsPerCoeff;
+                while (shift >= 8) {
+                    result[i++] = (byte) acc;
+                    acc >>= 8;
+                    shift -= 8;
+                }
+            }
+        }
```

Comment:
Add a comment here that `shift` must be zero now so we should have output all bits.